File: web/js/avatar_loader.js
/**
 * v7 Phase B — ReadyPlayerMe glb loader + Mixamo animation playback.
 *
 * Replaces the old `avatar_builder.js` (which built block-rigid meshes
 * out of Three.js primitives) with real game-quality glb models from
 * ReadyPlayerMe and skeletal animations from Mixamo.
 *
 * Public API:
 *
 *   const m = await loadAvatarManifest();              // {presets, poseMap}
 *   const avatar = await loadAvatar("male_casual_25"); // THREE.Group
 *   const mixer = playAnimation(avatar, "idle_relaxed");
 *   // ... in render loop: mixer.update(deltaSec)
 *
 * Falls back to the legacy procedural builder when a glb is missing,
 * so the 3D preview always renders something — even before the asset
 * pack is dropped into web/avatars/preset/.
 */

import * as THREE from "https://cdn.jsdelivr.net/npm/three@0.165.0/build/three.module.js";
import { GLTFLoader } from "three/addons/loaders/GLTFLoader.js";
// SkeletonUtils exports named helpers (clone, retargetClip, ...) rather
// than a SkeletonUtils namespace object — import them as a namespace.
import * as SkeletonUtils from "three/addons/utils/SkeletonUtils.js";

const ASSET_BASE = "/web/avatars";
const PRESET_DIR = `${ASSET_BASE}/preset`;
const BASE_AVATAR_URL = `${ASSET_BASE}/base/xbot.glb`;
const ANIM_DIR = `${ASSET_BASE}/animations`;
const FALLBACK_PRESETS = [
  {
    id: "female_youth_18",
    name_zh: "少女 · 18",
    gender: "female",
    age: 18,
    style: "youth",
    thumbnail: `${PRESET_DIR}/female_youth_18.png`,
    glb: `${PRESET_DIR}/female_youth_18.glb`,
  },
  {
    id: "male_casual_25",
    name_zh: "休闲男 · 25",
    gender: "male",
    age: 25,
    style: "casual",
    thumbnail: `${PRESET_DIR}/male_casual_25.png`,
    glb: `${PRESET_DIR}/male_casual_25.glb`,
  },
  {
    id: "female_casual_22",
    name_zh: "休闲女 · 22",
    gender: "female",
    age: 22,
    style: "casual",
    thumbnail: `${PRESET_DIR}/female_casual_22.png`,
    glb: `${PRESET_DIR}/female_casual_22.glb`,
  },
  {
    id: "female_elegant_30",
    name_zh: "优雅女 · 30",
    gender: "female",
    age: 30,
    style: "elegant",
    thumbnail: `${PRESET_DIR}/female_elegant_30.png`,
    glb: `${PRESET_DIR}/female_elegant_30.glb`,
  },
  {
    id: "female_artsy_25",
    name_zh: "文艺女 · 25",
    gender: "female",
    age: 25,
    style: "artsy",
    thumbnail: `${PRESET_DIR}/female_artsy_25.png`,
    glb: `${PRESET_DIR}/female_artsy_25.glb`,
  },
  {
    id: "male_business_35",
    name_zh: "商务男 · 35",
    gender: "male",
    age: 35,
    style: "business",
    thumbnail: `${PRESET_DIR}/male_business_35.png`,
    glb: `${PRESET_DIR}/male_business_35.glb`,
  },
  {
    id: "male_athletic_28",
    name_zh: "运动男 · 28",
    gender: "male",
    age: 28,
    style: "athletic",
    thumbnail: `${PRESET_DIR}/male_athletic_28.png`,
    glb: `${PRESET_DIR}/male_athletic_28.glb`,
  },
  {
    id: "child_boy_8",
    name_zh: "男孩 · 8",
    gender: "male",
    age: 8,
    style: "child",
    thumbnail: `${PRESET_DIR}/child_boy_8.png`,
    glb: `${PRESET_DIR}/child_boy_8.glb`,
  },
  {
    id: "child_girl_8",
    name_zh: "女孩 · 8",
    gender: "female",
    age: 8,
    style: "child",
    thumbnail: `${PRESET_DIR}/child_girl_8.png`,
    glb: `${PRESET_DIR}/child_girl_8.glb`,
  },
];

// ─────────────────────────────────────────────────────────────────────
// v7 Phase B+: 8 preset 色板
//
// Why this exists:
//   We didn't manage to wire up 8 distinct ReadyPlayerMe glb avatars
//   (every RPM avatar requires a manual create on readyplayer.me). To
//   still ship a "8 visually distinct, game-quality, rigged" gallery
//   we use Three.js's official Xbot.glb as a SHARED rigged humanoid
//   base mesh and apply a per-preset colour tint to the 2 PBR
//   materials Xbot exposes (the "joints" material and the "limbs"
//   material).
//
//   Xbot is a Khronos-published Mixamo-rigged sample model — same
//   skeleton bone names as anything you download from Mixamo, so all
//   30 placeholder Mixamo animation clips (and any future real
//   Mixamo download) can drive it without retargeting.
//
//   When the user upgrades to real RPM avatars later, replacing the
//   per-preset glbs in web/avatars/preset/<id>.glb takes precedence
//   automatically — see the load order in loadAvatar() below.
// ─────────────────────────────────────────────────────────────────────
const PRESET_TINTS = {
  male_casual_25:    { limbs: 0x4e8de0, joints: 0x33415f },  // skin-warm, navy tee
  male_business_35:  { limbs: 0x21263a, joints: 0x101321 },  // dark suit
  male_athletic_28:  { limbs: 0xd33b3b, joints: 0x6a1b1b },  // red sport top
  female_casual_22:  { limbs: 0xe89bb6, joints: 0x6a4452 },  // pink hoodie
  female_elegant_30: { limbs: 0xb33064, joints: 0x4d1730 },  // wine evening
  female_artsy_25:   { limbs: 0xc7b894, joints: 0x6a614e },  // beige bohemian
  child_boy_8:       { limbs: 0x5dbcdf, joints: 0x294f5d },  // sky blue
  child_girl_8:      { limbs: 0xf7b1d3, joints: 0x6c4356 },  // rose
};

// In-memory caches keyed by asset id. We deep-clone via SkeletonUtils
// every time we hand out an avatar so the same preset can show up on
// multiple shots (or multiple persons) at once without sharing state.
const avatarCache = new Map();
const animCache = new Map();
let manifestPromise = null;
let baseTemplate = null; // {scene, animations} - Xbot loaded once, shared
const loader = new GLTFLoader();

/**
 * Lazy-load and cache the shared Xbot base mesh (rigged humanoid with
 * Mixamo-standard skeleton + 7 built-in animations: idle / run /
 * agree / headShake / sad_pose / sneak_pose). Returned glTF is the
 * RAW template — callers should SkeletonUtils.clone() it.
 */
async function loadBaseTemplate() {
  if (baseTemplate) return baseTemplate;
  baseTemplate = (async () => {
    const tpl = await fetchGltf(BASE_AVATAR_URL);
    if (!tpl) {
      console.warn("[avatar_loader] Xbot base failed to load — gallery will fall back to procedural");
      return null;
    }
    return tpl;
  })();
  return baseTemplate;
}

/** Light-weight wrapper for the /avatars/manifest endpoint. */
export async function loadAvatarManifest({ baseUrl = "" } = {}) {
  if (manifestPromise) return manifestPromise;
  manifestPromise = (async () => {
    try {
      const res = await fetch(`${baseUrl}/avatars/manifest`);
      if (!res.ok) throw new Error(`manifest http ${res.status}`);
      const json = await res.json();
      const flat = flattenPoseMap(json.pose_to_mixamo || {});
      return {
        presets: mergePresetManifest(json.presets || []),
        poseMap: flat.flat,
        fallbackByCount: flat.fallbackByCount,
      };
    } catch (err) {
      console.warn("[avatar_loader] manifest fetch failed:", err);
      return {
        presets: mergePresetManifest([]),
        poseMap: {},
        fallbackByCount: {},
      };
    }
  })();
  return manifestPromise;
}

function mergePresetManifest(remotePresets) {
  const merged = [];
  const seen = new Set();
  for (const preset of [...remotePresets, ...FALLBACK_PRESETS]) {
    if (!preset?.id || seen.has(preset.id)) continue;
    seen.add(preset.id);
    merged.push({ ...preset, nameZh: preset.nameZh || preset.name_zh || preset.id });
  }
  return merged;
}

function flattenPoseMap(rawMap) {
  const flat = {};
  for (const section of ["single", "two_person", "three_person", "four_person"]) {
    Object.assign(flat, rawMap[section] || {});
  }
  return {
    flat,
    fallbackByCount: rawMap.fallback_by_count || { 1: "idle_relaxed" },
  };
}

/**
 * Fetch and cache a preset glb. Returns a deep clone so the caller can
 * place it in the scene without affecting later callers.
 *
 * Returns null on failure — caller should fall back to procedural mesh.
 *
 * @param {string} presetId  e.g. "male_casual_25"
 * @returns {Promise<THREE.Group | null>}
 */
export async function loadAvatar(presetId) {
  if (!presetId) return null;
  if (!avatarCache.has(presetId)) {
    avatarCache.set(presetId, fetchGltf(`${PRESET_DIR}/${presetId}.glb`));
  }
  const tpl = await avatarCache.get(presetId);
  if (!tpl) return null;
  // SkeletonUtils.clone preserves bones + animation tracks (the
  // difference that lets multi-instance animation work on real RPM
  // glb). For our placeholder glb (which has no SkinnedMesh / Skeleton)
  // SkeletonUtils.clone occasionally throws on certain three.js
  // versions; fall back to a plain deep clone in that case so the
  // procedural placeholder pipeline still upgrades to the real glb.
  try {
    return SkeletonUtils.clone(tpl);
  } catch (err) {
    console.debug("[avatar_loader] SkeletonUtils.clone failed, using plain clone:", err?.message || err);
    return tpl.clone(true);
  }
}

/**
 * Fetch and cache a Mixamo animation glb. Returns the AnimationClip
 * (the geometry is discarded — the clip retargets onto whichever
 * avatar skeleton is provided to playAnimation).
 *
 * @param {string} animId  e.g. "idle_relaxed"
 * @returns {Promise<THREE.AnimationClip | null>}
 */
export async function loadAnimationClip(animId) {
  if (!animId) return null;
  if (!animCache.has(animId)) {
    animCache.set(animId, fetchGltf(`${ANIM_DIR}/${animId}.glb`));
  }
  const tpl = await animCache.get(animId);
  if (!tpl || !tpl.animations || !tpl.animations.length) return null;
  return tpl.animations[0];
}

/**
 * Bind a Mixamo animation to an avatar and start playback.
 * Caller must:
 *   - keep the returned AnimationMixer
 *   - call mixer.update(deltaSec) every frame
 *
 * @returns {{mixer: THREE.AnimationMixer, action: THREE.AnimationAction} | null}
 */
export function playAnimation(avatar, clip, { loop = true } = {}) {
  if (!avatar || !clip) return null;
  const mixer = new THREE.AnimationMixer(avatar);
  const action = mixer.clipAction(clip);
  action.setLoop(loop ? THREE.LoopRepeat : THREE.LoopOnce, Infinity);
  if (!loop) action.clampWhenFinished = true;
  action.play();
  return { mixer, action };
}

/**
 * Pick the right preset id for a given person + shot context.
 *
 * Heuristic: respect the user's persisted preference first
 * (UserDefaults/sessionStorage 'avatarPicks'), then derive from the
 * pose's role (person_a / person_b...) using a simple alternation
 * rule so couples don't end up with two identical presets.
 */
export function mapAvatarPick(person, personIndex, presets) {
  // Read the persisted pick list using the SAME key the rest of the
  // app uses (store.js's `apc.avatarPicks`, localStorage scope). This
  // is what makes "pick once on the home page, see it everywhere"
  // behaviour work cross-page and (when the iOS app uses the same id
  // string) cross-device-class.
  const persisted = (() => {
    try {
      const raw = localStorage.getItem("apc.avatarPicks")
                || sessionStorage.getItem("apc.avatarPicks");
      return JSON.parse(raw || "[]");
    } catch { return []; }
  })();
  if (persisted[personIndex]) return persisted[personIndex];

  if (!presets || !presets.length) return null;
  // Default rotation: female lead → male partner → female friend → child
  const order = [
    "female_youth_18",
    "male_casual_25",
    "female_casual_22",
    "female_elegant_30",
    "child_girl_8",
  ];
  const orderById = order.filter((id) => presets.some((p) => p.id === id));
  return orderById[personIndex % orderById.length] || presets[0].id;
}

/**
 * Resolve the LLM-recommended pose id to a Mixamo animation id, with
 * person_count fallback. Mirrors backend lookup_mixamo_for_pose.
 */
export function resolveMixamoId(poseId, personCount, manifest) {
  if (!manifest) return "idle_relaxed";
  const direct = manifest.poseMap[poseId];
  if (direct) return direct;
  return manifest.fallbackByCount[String(personCount)] || "idle_relaxed";
}

// Internal — load + cache a glTF or null on failure.
function fetchGltf(url) {
  return new Promise((resolve) => {
    loader.load(
      url,
      (gltf) => resolve(gltf.scene ? Object.assign(gltf.scene, { animations: gltf.animations }) : null),
      undefined,
      (err) => {
        // We *expect* this to fail until the asset pack is shipped;
        // log only at debug level.
        console.debug("[avatar_loader] fetch failed:", url, err?.message || err);
        resolve(null);
      },
    );
  });
}

File: ios/AIPhotoCoach/Features/Avatar/AvatarLoader.swift
// v7 Phase B — RealityKit avatar + animation loader.
//
// Replaces AvatarBuilderSCN's hand-built SCNBox geometry with real
// USDZ models converted from ReadyPlayerMe glb files.
//
// Public API:
//
//   let m = try await AvatarManifest.shared.load()
//   let entity = try await AvatarLoader.shared.load(presetId: "male_casual_25")
//   try await AvatarLoader.shared.playAnimation("idle_relaxed", on: entity)
//
// All async; loads are cached in-memory so subsequent calls are
// instant. Falls back to nil + a console warning when the .usdz file
// isn't bundled — caller should detect and fall back to the legacy
// SceneKit avatar.

import Foundation
import RealityKit

/// One preset avatar's metadata, mirrors the backend manifest entry.
public struct AvatarPresetEntry: Codable, Hashable, Sendable {
    public let id: String
    public let nameZh: String
    public let gender: String
    public let age: Int
    public let style: String
    public let tags: [String]
    public let glb: String       // web URL — used by Web only
    public let usdz: String      // bundle-relative path
    public let thumbnail: String

    enum CodingKeys: String, CodingKey {
        case id, gender, age, style, tags, glb, usdz, thumbnail
        case nameZh = "name_zh"
    }
}

/// A single Mixamo animation reference. We store the path relative to
/// the bundle's resource directory so RealityKit can load it directly.
public struct AvatarAnimationManifest: Codable, Sendable {
    public let single: [String: String]
    public let twoPerson: [String: String]
    public let threePerson: [String: String]
    public let fourPerson: [String: String]
    public let fallbackByCount: [String: String]
    /// B-pose-for-height — map from HeightHint enum string ("low",
    /// "eye_level", "high", "overhead") to a mixamo-id whose stance
    /// makes sense at that altitude. When the manifest omits this map
    /// we fall back to ``resolve(poseId:personCount:)`` so older
    /// backend versions keep working without any null checks here.
    public let poseForHeight: [String: String]?

    enum CodingKeys: String, CodingKey {
        case single
        case twoPerson = "two_person"
        case threePerson = "three_person"
        case fourPerson = "four_person"
        case fallbackByCount = "fallback_by_count"
        case poseForHeight = "pose_for_height"
    }

    /// Flatten all sections into a single pose-id → mixamo-id map.
    public var flatPoseMap: [String: String] {
        var combined: [String: String] = [:]
        for section in [single, twoPerson, threePerson, fourPerson] {
            combined.merge(section) { _, new in new }
        }
        return combined
    }

    public func resolve(poseId: String?, personCount: Int) -> String {
        if let id = poseId, let direct = flatPoseMap[id] {
            return direct
        }
        return fallbackByCount[String(personCount)] ?? "idle_relaxed"
    }

    /// Pick a pose specifically tailored to the subject's vertical
    /// position. ``heightHint`` strings match the backend HeightHint
    /// enum. Returns a sensible built-in fallback when the manifest
    /// doesn't carry ``pose_for_height``.
    public func resolveForHeight(_ heightHint: String?) -> String? {
        guard let h = heightHint else { return nil }
        if let mapped = poseForHeight?[h] { return mapped }
        // Built-in defaults so "overhead" doesn't fall back to a
        // standing idle, which would look wrong floating 3 m in the air.
        switch h {
        case "overhead":  return "lean_look_down"
        case "high":      return "stand_look_down_slight"
        case "low":       return "look_up_curious"
        default:          return nil
        }
    }
}

/// Container for the /avatars/manifest payload.
public struct AvatarManifestPayload: Codable, Sendable {
    public let presets: [AvatarPresetEntry]
    public let poseToMixamo: AvatarAnimationManifest

    enum CodingKeys: String, CodingKey {
        case presets
        case poseToMixamo = "pose_to_mixamo"
    }
}

/// Caches the manifest fetched from the backend. The avatar gallery
/// + AR view both subscribe to this to learn which presets are
/// available.
@MainActor
public final class AvatarManifest: ObservableObject {
    public static let shared = AvatarManifest()

    @Published public private(set) var payload: AvatarManifestPayload?
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var lastError: String?

    private init() {}

    /// Fetch + cache. Subsequent calls return the cached payload
    /// without going to the network.
    @discardableResult
    public func load(baseURL: URL? = nil) async -> AvatarManifestPayload? {
        if let payload { return payload }
        if isLoading { return nil }
        isLoading = true
        defer { isLoading = false }
        let base = baseURL ?? AvatarManifest.defaultBaseURL()
        guard let url = URL(string: "/avatars/manifest", relativeTo: base) else {
            lastError = "invalid base URL"
            return nil
        }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let decoder = JSONDecoder()
            let parsed = try decoder.decode(AvatarManifestPayload.self, from: data)
            let merged = AvatarManifestPayload(
                presets: Self.mergeBundledPresets(parsed.presets),
                poseToMixamo: parsed.poseToMixamo
            )
            self.payload = merged
            return merged
        } catch {
            lastError = error.localizedDescription
            print("[AvatarManifest] load failed:", error)
            let fallback = AvatarManifestPayload(
                presets: Self.mergeBundledPresets([]),
                poseToMixamo: AvatarAnimationManifest(
                    single: [:],
                    twoPerson: [:],
                    threePerson: [:],
                    fourPerson: [:],
                    fallbackByCount: ["1": "idle_relaxed"],
                    poseForHeight: nil
                )
            )
            self.payload = fallback
            return fallback
        }
    }

    private static func mergeBundledPresets(_ remote: [AvatarPresetEntry]) -> [AvatarPresetEntry] {
        var merged: [AvatarPresetEntry] = []
        var seen = Set<String>()
        for preset in remote + bundledPresetFallbacks() {
            guard !seen.contains(preset.id) else { continue }
            seen.insert(preset.id)
            merged.append(preset)
        }
        return merged
    }

    private static func bundledPresetFallbacks() -> [AvatarPresetEntry] {
        [
            AvatarPresetEntry(id: "female_youth_18", nameZh: "少女 · 18", gender: "female", age: 18, style: "youth", tags: ["youth", "summer", "sweet"], glb: "/web/avatars/preset/female_youth_18.glb", usdz: "Avatars/female_youth_18.usdz", thumbnail: "/web/avatars/preset/female_youth_18.png"),
            AvatarPresetEntry(id: "male_casual_25", nameZh: "休闲男 · 25", gender: "male", age: 25, style: "casual", tags: ["street", "everyday"], glb: "/web/avatars/preset/male_casual_25.glb", usdz: "Avatars/male_casual_25.usdz", thumbnail: "/web/avatars/preset/male_casual_25.png"),
            AvatarPresetEntry(id: "female_casual_22", nameZh: "休闲女 · 22", gender: "female", age: 22, style: "casual", tags: ["youth", "street"], glb: "/web/avatars/preset/female_casual_22.glb", usdz: "Avatars/female_casual_22.usdz", thumbnail: "/web/avatars/preset/female_casual_22.png"),
            AvatarPresetEntry(id: "female_elegant_30", nameZh: "优雅女 · 30", gender: "female", age: 30, style: "elegant", tags: ["formal", "fashion"], glb: "/web/avatars/preset/female_elegant_30.glb", usdz: "Avatars/female_elegant_30.usdz", thumbnail: "/web/avatars/preset/female_elegant_30.png"),
            AvatarPresetEntry(id: "female_artsy_25", nameZh: "文艺女 · 25", gender: "female", age: 25, style: "artsy", tags: ["bohemian", "softlight"], glb: "/web/avatars/preset/female_artsy_25.glb", usdz: "Avatars/female_artsy_25.usdz", thumbnail: "/web/avatars/preset/female_artsy_25.png"),
            AvatarPresetEntry(id: "male_business_35", nameZh: "商务男 · 35", gender: "male", age: 35, style: "business", tags: ["formal", "office"], glb: "/web/avatars/preset/male_business_35.glb", usdz: "Avatars/male_business_35.usdz", thumbnail: "/web/avatars/preset/male_business_35.png"),
            AvatarPresetEntry(id: "male_athletic_28", nameZh: "运动男 · 28", gender: "male", age: 28, style: "athletic", tags: ["outdoor", "fit"], glb: "/web/avatars/preset/male_athletic_28.glb", usdz: "Avatars/male_athletic_28.usdz", thumbnail: "/web/avatars/preset/male_athletic_28.png"),
            AvatarPresetEntry(id: "child_boy_8", nameZh: "男孩 · 8", gender: "male", age: 8, style: "child", tags: ["family", "kids"], glb: "/web/avatars/preset/child_boy_8.glb", usdz: "Avatars/child_boy_8.usdz", thumbnail: "/web/avatars/preset/child_boy_8.png"),
            AvatarPresetEntry(id: "child_girl_8", nameZh: "女孩 · 8", gender: "female", age: 8, style: "child", tags: ["family", "kids"], glb: "/web/avatars/preset/child_girl_8.glb", usdz: "Avatars/child_girl_8.usdz", thumbnail: "/web/avatars/preset/child_girl_8.png"),
        ]
    }

    /// Resolve the API base URL from UserDefaults (BYOK / dev override)
    /// or fall back to localhost.
    private static func defaultBaseURL() -> URL {
        if let saved = UserDefaults.standard.string(forKey: "apiBaseURL"),
           let u = URL(string: saved) {
            return u
        }
        return URL(string: "http://127.0.0.1:8000")!
    }
}

// MARK: - Loader

/// Loads + caches RealityKit Entities from bundled USDZ avatars.
@MainActor
public final class AvatarLoader {
    public static let shared = AvatarLoader()

    private var avatarCache: [String: Entity] = [:]
    private var animationCache: [String: Entity] = [:]

    private init() {}

    /// Load (or return cached) avatar template by preset id. The
    /// returned Entity is *cloned* so each call gives a fresh instance
    /// callers can mutate / animate independently.
    public func load(presetId: String) async throws -> Entity? {
        if let cached = avatarCache[presetId] {
            return cached.clone(recursive: true)
        }
        let candidate = "Avatars/\(presetId)"
        do {
            // Entity(named:in:) async API is iOS 18+. On iOS 17 the
            // synchronous loader can throw via the same name; this
            // gate makes the call site explicitly 18-only and lets
            // older devices fall through to the SCN fallback path
            // that ARGuideView already implements.
            let entity: Entity
            if #available(iOS 18.0, *) {
                entity = try await Entity(named: candidate, in: nil)
            } else {
                entity = try Entity.load(named: candidate)
            }
            avatarCache[presetId] = entity
            return entity.clone(recursive: true)
        } catch {
            print("[AvatarLoader] preset \(presetId) not bundled — falling back to legacy SCN. Reason:", error)
            return nil
        }
    }

    /// Load (or return cached) animation by id. Animations are USDZ
    /// files containing only the AnimationLibrary track — they're
    /// retargeted onto whichever avatar Entity is passed.
    public func loadAnimation(_ animId: String) async throws -> AnimationResource? {
        if let cached = animationCache[animId],
           let res = cached.availableAnimations.first {
            return res
        }
        let candidate = "Animations/\(animId)"
        do {
            let entity: Entity
            if #available(iOS 18.0, *) {
                entity = try await Entity(named: candidate, in: nil)
            } else {
                entity = try Entity.load(named: candidate)
            }
            animationCache[animId] = entity
            return entity.availableAnimations.first
        } catch {
            print("[AvatarLoader] animation \(animId) not bundled. Reason:", error)
            return nil
        }
    }

    /// Convenience — resolve LLM pose id → Mixamo id → load + bind.
    /// Returns the running animation controller if everything aligned.
    @discardableResult
    public func playPose(
        _ poseId: String?,
        personCount: Int,
        on entity: Entity,
        manifest: AvatarAnimationManifest?,
    ) async -> AnimationPlaybackController? {
        let animId = manifest?.resolve(poseId: poseId, personCount: personCount)
            ?? "idle_relaxed"
        guard let animation = try? await loadAnimation(animId) else { return nil }
        return entity.playAnimation(animation.repeat(), transitionDuration: 0.4)
    }
}

// MARK: - Pick helpers

public enum AvatarPicker {
    /// Map a person index in a multi-person shot to a preset id, using
    /// the user's persisted picks where available, then a sensible
    /// rotation through the pack so couples aren't twins.
    public static func pick(
        personIndex: Int,
        from presets: [AvatarPresetEntry],
        persisted: [String]? = nil,
    ) -> String? {
        let stored = persisted ?? UserDefaults.standard.stringArray(forKey: "avatarPicks") ?? []
        if personIndex < stored.count, !stored[personIndex].isEmpty {
            return stored[personIndex]
        }
        guard !presets.isEmpty else { return nil }
        let rotation = ["female_youth_18", "male_casual_25",
                        "female_casual_22",
                        "female_elegant_30", "child_girl_8"]
        let available = rotation.filter { id in presets.contains { $0.id == id } }
        if available.isEmpty {
            return presets[personIndex % presets.count].id
        }
        return available[personIndex % available.count]
    }
}
